reference/src/representation/enums.md

```diff
@@ -0,0 +1,376 @@
+# Representation of Rust `enum` types
+
+**Disclaimer:** Some parts of this section were decided in RFCs, but
+others represent the consensus from issue [#10]. The text will attempt
+to clarify which parts are "guaranteed" (owing to the RFC decision)
+and which parts are still in a "preliminary" state.
+
+[#10]: https://github.com/rust-rfcs/unsafe-code-guidelines/issues/10
+
+## Background
+
+**C-like enums.** The simplest form of enum is simply a list of
+variants:
+
+```rust
+enum SomeEnum {
+  Variant1,
+  Variant2,
+  Variant3,
+```
+
+Such enums are called "C-like" because they correspond quite closely
+with enums in the C language (though there are important differences
+as well, covered later). Presuming that they have more than one
+variant, these sorts of enums are always represented as a simple integer,
+though the size will vary.
+
+C-like enums may also specify the value of their discriminants explicitly:
+
+```rust
+enum SomeEnum {
+  Variant22 = 22,
+  Variant44 = 44,
+  Variant45,
+}
+```
+
+As in C, discriminant values that are not specified are defined as
+either 0 (for the first variant) or as one more than the prior
+variant.
+
+**Data-carrying enums.** Enums whose enums have fields are called
+"data-carrying" enums. Note that for the purposes of this definition,
+it is not relevant whether those fields are zero-sized. Therefore this
+enum is considered "data-carrying":
+
+```rust
+enum Foo {
+  Bar(()),
+  Baz,
+}
+```
+
+**Option-like enums.** As a special case of data-carrying enums, we
+identify "option-like" enums as enums where all of the variants but
+one have no fields, and one variant has a single field. The most
+common example is `Option` itself. In some cases, as described below,
+the compiler may apply special optimization rules to the layout of
+option-like enums. The **payload** of an option-like enum is the value
+of that single field.
+
+## Enums with a specified representation
+
+Enums may be annotation using the following `#[repr]` tags:
+
+- A specific integer type (called `Int` as a shorthand below):
+  - `#[repr(u8)]`
+  - `#[repr(u16)]`
+  - `#[repr(u32)]`
+  - `#[repr(u64)]`
+  - `#[repr(i8)]`
+  - `#[repr(i16)]`
+  - `#[repr(i32)]`
+  - `#[repr(i64)]`
+- C-compatible layout:
+  - `#[repr(C)]`
+- C-compatible layout with a specified discriminant size:
+  - `#[repr(C, u8)]`
+  - `#[repr(C, u16)]`
+  - etc
+
+We cover each of the categories below. The layout rules for enums with
+explicit `#[repr]` annotations are specified in [RFC 2195][].
+
+[RFC 2195]: https://rust-lang.github.io/rfcs/2195-really-tagged-unions.html
+
+### Layout of an enum with no variants
+
+An enum with no variants can never be instantiated and is logically
+equivalent to the "never type" `!`. Such enums are guaranteed to have
+the same layout as `!` (zero size and alignment 1).
```

Comment:
I was surprised to find out that `#[repr(align(N))] enum Foo { ... }` doesn't work. Might be worth mentioning, otherwise future readers might (like me) ask themselves whether they can't set alignment higher.